upstairs/src/lib.rs

```diff
@@ -818,34 +936,48 @@ impl Work {
         ds_id: u64,
         client_id: u8,
         read_data: Option<Bytes>,
+        result: Result<(), CrucibleError>,
     ) -> Result<bool> {
         /*
          * Assume we don't have enough completed jobs, and only change
          * it if we have the exact amount required
          */
         let mut notify_guest = false;
+
         /*
-         * Get the completed count now (and add one for ourselves),
-         * because the job self ref won't let us call stat_count once we are
+         * Get the completed count now,
+         * because the job self ref won't let us call state_count once we are
          * using that ref, and the number won't change while we are in
          * this method (you did get the lock first, right??).
          */
         let wc = self.state_count(ds_id)?;
-        let jobs_completed = 1 + wc.done + wc.acked + wc.ack_ready;
+        let mut jobs_completed_ok = wc.completed_ok();
 
         let job = self
             .active
             .get_mut(&ds_id)
             .ok_or_else(|| anyhow!("reqid {} is not active", ds_id))?;
 
-        let oldstate = job.state.insert(client_id, IOState::Done);
-        assert_ne!(oldstate, Some(IOState::Done));
+        let newstate = if let Err(e) = result {
+            IOState::Error(e)
+        } else {
+            jobs_completed_ok += 1;
+            IOState::Done
+        };
+
+        let oldstate = job.state.insert(client_id, newstate);
+
+        if let Some(oldstate) = oldstate {
+            // we shouldn't be transitioning a state that was already transitioned
+            assert!(matches!(oldstate, IOState::New | IOState::InProgress));
+        } else {
+            panic!("no old state! that's bad!");
+        }
 
         /*
-         * XXX Handle errors here.
-         * Any error needs to be passed to Nexus (probably)
-         * 2 errors for Write/Flush,  3 errors for Reads
-         * should get passed back to the guest.
+         * Transition this job from Done to AckReady if enough have returned ok.
+         *
+         * XXX Any error needs to be passed to Nexus
```

Comment:
Any write or flush error is also serious, as we have to take some action on it as this downstairs is (possibly) no longer the same as the other two downstairs.

Once an error like this happens to one of the downstairs, we can't send reads to that same offset-range until we have repaired the data.  We also don't currently have a mechanism to prevent that, so we are opening the door to returning incorrect data until this downstairs is repaired.

If we can't think of something easy to do in this review, then very quickly afterwards we should put in the "block this downstairs from reads" switch we can throw to at least prevent silently returning the old data.  This also assumes the case where a write/flush fails, but a read would succeed.

There is a fair amount of work around recovery, and It's not yet time to start the real solution here, but we should do something to prevent bad data being returned and messing up some other unrelated test and us spending time trying to figure out what happened :)

Reply:
I tried something in 058c9bb:
> Transition downstairs to Failed if they return errors
>
> Keep track of downstairs errors in Work, and mark downstairs as bad if
> they return errors. Transition downstairs to Failed and skip sending
> them work if they are marked as bad.